numpyro/contrib/control_flow/scan.py

```diff
@@ -194,7 +270,14 @@ def g(*args, **kwargs):
         msg = apply_stack(initial_msg)
         (length, rng_key, carry), (pytree_trace, ys) = msg['value']
 
-    for msg in pytree_trace.trace.values():
-        apply_stack(msg)
+    if not msg["kwargs"].get("enum", False):
+        for msg in pytree_trace.trace.values():
+            apply_stack(msg)
+    else:
+        for msg in pytree_trace.trace.values():
+            with LocalNamedMessenger():
+                dim_to_name = msg["infer"].get("dim_to_name")
+                to_funsor(msg["value"], dim_to_name=OrderedDict([(k, dim_to_name[k]) for k in sorted(dim_to_name)]))
+                apply_stack(msg)
```

Comment:
This logic is fine, the `LocalNamedMessenger` here ensures that these dimensions have no effect on the rest of model execution.